### DISCUSSION TITLE
getSession is undefined

### DISCUSSION THREAD

**Discussion by lunawen** [enisdenjo/graphql-ws#352] [Q&A] [answered]
Hi, I'm new to GraphQL and today I'm using this library to learn subscription.

I followed the official doc and added graphql-ws to my project, with the example code.

However, I keep getting the "Unhandled GraphQL subscription error ReferenceError: getSession is not defined" error on my console.

Is there another library that I need to import? I googled but couldn't find it anywhere.. It would be good if the documentation can be clearer.

![Snipaste_2022-04-16_18-38-26](https://user-images.githubusercontent.com/34984218/163668521-8bd3c893-85b9-45ca-b661-5a807679a5af.png)
![Snipaste_2022-04-16_18-37-56](https://user-images.githubusercontent.com/34984218/163668525-28c60127-8ae1-49a3-8673-8bb4b8d4d011.png)

Thanks for your help!

**Comment by enisdenjo** [comment, marked as answer]
`getSession` is a placeholder example function that gets _some_ session information meant to be implemented by _you_, it doesn't exist in this library.

**Comment by lunawen** [comment]
Ahhh, thanks for the clarification.

Luna

**Reply by enisdenjo** [reply]
If you have suggestions on how to make this clearer in the example, I'd love to hear it! PRs are always welcome. 😄

**Reply by paymog** [reply]
It would be wonderful if there was an example with how we can get session information in apollo client. I'm also having some trouble figuring how to to get access to cookies which hold a JWT that I want to pass to the backend for authentication of the websocket.

**Reply by enisdenjo** [reply]
Cookies are automatically added by the browser for each request, including the WebSocket upgrade request (of course, if [SameSite](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite) attribute is set up correctly) - there's nothing you have to do on the client.

If using `useServer from 'graphql-ws/lib/use/ws'`, the [upgrade request is available in the context extras](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/use_ws.Extra.md#request).
```ts
...

useServer(
  {
    ...
    onConnect: (ctx) => {
      ctx.extra.request.headers.cookie; // <- your Cookie header
    },
    onSubscribe: (ctx) => {
      ctx.extra.request.headers.cookie; // <- your Cookie header
    },
    onNext: (ctx) => {
      ctx.extra.request.headers.cookie; // <- your Cookie header
    },
    onError: (ctx) => {
      ctx.extra.request.headers.cookie; // <- your Cookie header
    },
    onComplete: (ctx) => {
      ctx.extra.request.headers.cookie; // <- your Cookie header
    },
  },
  ...
);

```

**Reply by paymog** [reply]
ah, it looks like the cookies are included automatically serverside when using `useServer` under `ctx.extra.request.rawHeaders`. It seems the cookies need to be manually extracted from the raw headers though. 

Something like this seems to be working for me:

```
      context: async (ctx, msg, args) => {
        // Returning an object will add that information to our
        // GraphQL context, which all of our resolvers have access to.
        const context = {};

        const rawHeaders = ctx.extra.request.rawHeaders;
        const cookieIndex = rawHeaders.indexOf('Cookie');
        const rawCookies = rawHeaders[cookieIndex + 1];
        const cookies = rawCookies.split(';');

        for (let i = 0; i < cookies.length; i++) {
          const cookie = cookies[i];
          const [key, value] = cookie.trim().split('=');
          if (key === 'app_login_token') {
            context.user = await getUserWithToken({ token: value });
            break;
          }
        }

        return context;
      },
```

**Reply by paymog** [reply]
ah, just saw your message while I was writing mine! Is it better to handle jwt auth over cookies in the various `on...` methods above or to set up the context like I've shown?

**Reply by enisdenjo** [reply]
I recommend, and personally always prefer, using [`onConnect` callback](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#onconnect) for authenticating users and `context` for injecting into GraphQL's context. Roughly (warning, untested code):
```ts
import { parse as parseCookie } from 'cookie'; // yarn add cookie

...

useServer(
  {
    ...
    onConnect: async (ctx) => {
      const cookies = parseCookie(ctx.extra.request.headers.cookie);
      const authenticatedUser = await getUserWithToken({ token: cookies.app_login_token });
      if (!authenticatedUser) {
        return false; // will reject connection and close the WebSocket
      }

      // store the authenticated user in the context extra for easy reuse
      ctx.extra.user = authenticatedUser;
    },
    context: (ctx) => ({
      // inject the context's authenticated user into GraphQL context
      user: ctx.extra.user,
    }),
  },
  ...
);
```

P.S. you're better off using the `headers.cookies` over `rawHeaders`, NodeJS does the parsing from https://github.com/enisdenjo/graphql-ws/discussions/352#discussioncomment-2673054 for you.

**Reply by paymog** [reply]
ah yup, using `headers.cookies` is *much* nicer! 

Is there a particular reason you recommend and prefer using `onConnect` for auth? I have an intuition for why it's better but would like to know your reasons if you don't mind sharing.

**Reply by enisdenjo** [reply]
>  Is the connection callback called when the client requests the connection initialisation through the message `ConnectionInit`.
> 
> The message payload (`connectionParams` from the client) is present in the `Context.connectionParams`.
> 
> - Returning `true` or nothing from the callback will allow the client to connect.
> 
> - Returning `false` from the callback will terminate the socket by dispatching the  close event `4403: Forbidden`.
> 
> - Returning a `Record` from the callback will allow the client to connect and pass the returned value to the client through the optional `payload` field in the `ConnectionAck` message.

_[From `onConnect` documentation.](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#onconnect)_

In essence, it's the best place because it is the "connection request" callback - the _first_ thing that gets called when a client is establishing a connection.

**Comment by lunawen** [comment]
Sure, I'd love to! Just created a PR with some comments :)

Luna
